src/opm/parser/eclipse/EclipseState/Grid/FieldProps.cpp

```diff
@@ -596,6 +596,25 @@ void FieldProps::erase<double>(const std::string& keyword) {
     this->double_data.erase(keyword);
 }
 
+template <>
+std::vector<int> FieldProps::extract<int>(const std::string& keyword) {
+    auto field_iter = this->int_data.find(keyword);
+    auto field = std::move(field_iter->second);
+    std::vector<int> data = std::move( field.data );
+    this->int_data.erase( field_iter );
+    return data;
+}
+
+template <>
+std::vector<double> FieldProps::extract<double>(const std::string& keyword) {
+    auto field_iter = this->double_data.find(keyword);
+    auto field = std::move(field_iter->second);
```

Comment:
This looks weired. You seem to be moving the data out of the container `double_data`. Doesn't that mean that `double_data[keyword]` will afterwards be in undefined state? I would expect this to cause some side effects when calling
this function again. What am I missing?

Reply:
Yes - I am moving the data out of the container and returning it; but the data is also explicitly removed from the container with `double_datae.erase( iter )` - so it should not be undefined.

Reply:
??? Hence the a second call to get_copy will, throw? That might come as a surprise. If read get_copy I would think that the origin of the copy stays untouched.

Reply:
No; we are now discussing around the private `::extract<T>( )` method - it is probably better to look at the `FieldProps::get_copy<T>` to understand the semantics:
```C+
std::vector<T> get_copy(const std::string& keyword, bool global) {
        if (this->has<T>(keyword)) {
            const auto& data = this->get_valid_data<T>(keyword);

            if (global)
                return this->global_copy(data);
            else
                return data;
        } else {
            const auto& field_ptr = this->try_get<T>(keyword);
            if (!field_ptr)
                throw std::invalid_argument("No such valid keyword: " + keyword);

            if (global)
                return this->global_copy(this->extract<T>(keyword));
            else
                return this->extract<T>(keyword);
        }
    }

```
i.e. the `extract` method is only used to return data which has been auto-created by the `try_get` call - which was not there originally. I.e.

 - If the container contains `keyword` - just return the data.
 - If the container does not contain keyword:
   1. Let the machinery create and initialize the keyword.
   2. Move the newly created data out of the container.
   3. Erase the (moved out of) data from the container.
   4. Return the data.